src/AppInstallerCommonCore/Settings.cpp

```diff
@@ -128,66 +111,143 @@ namespace AppInstaller::Settings
+            void EnsureParentPath()
             {
-                std::filesystem::path result = m_root;
+                std::filesystem::create_directories(m_settingFile.parent_path());
+            }
 
-                if (name.has_parent_path())
+            std::filesystem::path m_settingFile;
+        };
+
+        // A settings container that manages safely writing to its value with exchange semantics.
+        // Only allows Set to succeed if the hash value of the setting is the same as the last time it was read.
+        struct ExchangeSettingsContainer : public details::ISettingsContainer
+        {
+            ExchangeSettingsContainer(std::unique_ptr<ISettingsContainer>&& container, const std::string_view& name) :
+                m_container(std::move(container)), m_name(name) {}
+
+            std::unique_ptr<std::istream> Get() override
+            {
+                return GetInternal(m_hash);
+            }
+
+            bool Set(std::string_view value) override
+            {
+                THROW_HR_IF(E_UNEXPECTED, value.size() > std::numeric_limits<uint32_t>::max());
+
+                // If Set is called without ever reading the value, then we can assume that caller wants
+                // to overwrite it regardless. Also, we don't have any previous value to compare against
+                // anyway so the only other option would be to always reject it.
+                if (m_hash)
                 {
-                    result /= name.parent_path();
+                    std::optional<SHA256::HashBuffer> currentHash;
+                    std::ignore = GetInternal(currentHash);
+
+                    if (currentHash && !AreHashesEqual(m_hash.value(), currentHash.value()))
+                    {
+                        AICLI_LOG(Core, Verbose, << "Setting value for '" << m_name << "' has changed since last read; rejecting Set");
+                        return false;
+                    }
                 }
 
-                if (createParent)
+                SHA256::HashBuffer newHash = SHA256::ComputeHash(reinterpret_cast<const uint8_t*>(value.data()), static_cast<uint32_t>(value.size()));
+                if (m_container->Set(value))
                 {
-                    std::filesystem::create_directories(result);
+                    m_hash = std::move(newHash);
+                    return true;
                 }
+                else
+                {
+                    return false;
+                }
+            }
 
-                result /= name.filename();
-                return result;
+            void Remove() override
+            {
+                m_container->Remove();
+                m_hash.reset();
+            }
+
+            std::filesystem::path PathTo() override
+            {
+                return m_container->PathTo();
+            }
+
+        protected:
+            static bool AreHashesEqual(const SHA256::HashBuffer& a, const SHA256::HashBuffer& b)

```

Comment:
>static bool AreHashesEqual(const SHA256::HashBuffer& a, const SHA256::HashBuffer& b) [](http://example.com/codeflow?start=12&length=84)

nit: this could be in SHA256.h so other code can use this as well